elasticapm/utils/time.py

```diff
@@ -0,0 +1,116 @@
+    def __init__(self, src_timefn, dst_timefn, measurements_ttl=None):
+        self.__lock = threading.Lock()
+        self.__src_timefn = src_timefn
+        self.__dst_timefn = dst_timefn
+        self.__measurements_ttl = measurements_ttl
+        self.__clocks_measurement_cache = None
+
+    def take_clock_measurements(
+        self,
+        tolerance=0.00001,
+        limit=10
+    ):
+        """Attempt to take times from both clocks as close as possible."""
+        if limit <= 0:
+            raise ValueError(
+                "'limit' can not be less than 1. Got {}.".format(limit)
+            )
+        src_now = dst_now = None
+        epsilon = sys.float_info.max
+        i = 0
+        for i in range(limit):
+            src_before = self.__src_timefn()
+            dst_between = self.__dst_timefn()
+            src_after = self.__src_timefn()
+            src_diff = src_after - src_before
+            delta = abs(src_diff)
+            if delta < epsilon:
+                src_now = src_before + src_diff / 2
+                dst_now = dst_between
+                epsilon = delta
+            if epsilon <= tolerance:
+                break
+        else:
+            raise ClockMeasurementsFailed(
+                'Failed to take clock measurements in {} iteration.'.format(i)
+            )
+        measurements = ClockMeasurements(src_now, dst_now)
+        if self.__measurements_ttl:
+            with self.__lock:
+                self.__clocks_measurement_cache = (
+                    time.time(), tolerance, measurements
+                )
+
+        return measurements
+
+    def __check_measurements_cache(self, tolerance):
+        """Check if there are cached clock measurements."""
+        if self.__clocks_measurement_cache is None:
+            return
+        if self.__measurements_ttl is None or self.__measurements_ttl == 0:
+            return
+        if ((self.__clocks_measurement_cache[0] + self.__measurements_ttl) >
+            time.time()):
+            return
+        if self.__clocks_measurement_cache[1] != tolerance:
+            return
+        return self.__clocks_measurement_cache[2]
+
+    def convert_timestamp(
+        self,
+        timestamp,
+        clocks_measurements=None,
+        tolerance=0.00001,
+        limit=10
+    ):
+        """Convert timestamp between clocks representations."""
+        if not clocks_measurements:
+            with self.__lock:
+                clocks_measurements = self.__check_measurements_cache(
+                    tolerance=tolerance
+                )
+
+        if not clocks_measurements:
+            clocks_measurements = self.take_clock_measurements(
+                tolerance=tolerance,
+                limit=limit
+            )
+        return (
+            clocks_measurements.destination +
+            (timestamp - clocks_measurements.source)
+        )
+
+
+SYSTEM_TO_MONOTONIC = TimestampConversionUtil(
+    time.time,
+    time.monotonic,
```

Comment:
`time.monotonic` is not available on Python 2.7. To ensure Python 2.7 compatibility, `timeit.default_timer` is the best option. On Python 3, it defaults to `time.perf_counter`, on Python 2 to the best platform dependent time source (non-monotonic)